### PR TITLE
exporting normalizePathEnd

### PR DESCRIPTION
**Pull request by disruptek**
I dunno if this is the best way to conditionally export this proc in 1.1.

### CODE REVIEW

**Comment by timotheecour** [comment]
> I dunno if this is the best way to conditionally export this proc in 1.1.

seems like the best way to me

**Comment by disruptek** [comment]
What do you think about renaming the copy version to `normalizedPathEnd`?

**Comment by timotheecour** [comment]
my 2 cents: I know there's a precedent for that (eg: `sort` vs `sorted`) but:
* it's not universally followed
* nim doesn't auto-discard (best language decision never) so there's no ambiguity in general (except some corner cases) in using same name for both versions
* outparam(https://github.com/nim-lang/Nim/pull/11916) or sugar.outplace would also use same name

File: lib/pure/os.nim [hunk 1]
```diff
@@ -103,9 +106,24 @@ proc normalizePathEnd(path: var string, trailingSep = false) =
...
+      assert normalizePathEnd("/lib/", trailingSep = true) == "/lib/"
+      assert normalizePathEnd("", trailingSep = true) == "/"
+      assert normalizePathEnd("/lib", trailingSep = false) == "/lib"
+      assert normalizePathEnd("lib//", trailingSep = false) == "lib"
+      assert normalizePathEnd("lib", trailingSep = false) == "lib"
+      assert normalizePathEnd("/lib//", trailingSep = true) == "/lib/"
   result = path
   result.normalizePathEnd(trailingSep)
 
+when (NimMajor, NimMinor) >= (1, 1):
```
Comment (juancarlospaco) [inline lib/pure/os.nim:101]: This is not what `since` template is for?. :thinking:

Comment (disruptek) [reply lib/pure/os.nim:101]: The problem is that these procs already existed; we're just exporting them as of 1.1, but not before.

Comment (timotheecour) [reply lib/pure/os.nim:101]: That doesnt matter (implementation detail), what matters is public api

**Comment by disruptek** [comment]
Divergent symbols are easier to recognize in IDE completion and easier to control in import statements, but if no one wants to rename it, I'm fine with it as is.

**Comment by Araq** [comment]
The problem with this PR is that technically speaking you're relying on a compiler bug to achieve this. ``export`` is for exporting symbols **from an imported module**.

**Comment by timotheecour** [comment]
> The problem with this PR is that technically speaking you're relying on a compiler bug to achieve this. export is for exporting symbols from an imported module.

the "compiler bug" is IMO the more useful, consistent, less surprising behavior (and I've written code relying on that); we can just update the spec accordingly. The use-case in this PR is a good example; if we followed current spec, it'd lead to less DRY code.

**Comment by disruptek** [comment]
Hence my first comment.  How would you like to do it?

**Comment by Araq** [comment]
> the "compiler bug" is IMO the more useful, consistent, less surprising behavior (and I've written code relying on that); we can just update the spec accordingly. 

Agreed, however it then needs to have tests of its own to ensure we don't break this feature.

**Comment by timotheecour** [comment]
> And this also includes little things like "nim doc needs to list every exported proc..."

```nim
proc main()=discard
export main
```
just checked, `nim doc` indeed needs to be fixed to list `main` as public symbol instead of adding `main` in exports section. hopefully not hard

File: lib/pure/os.nim [hunk 2]
```diff
@@ -102,9 +98,59 @@ proc normalizePathEnd(path: var string, trailingSep = false) =
...
+        assert normalizePathEnd("/lib", trailingSep = false) == "/lib"
+        assert normalizePathEnd("lib//", trailingSep = false) == "lib"
+        assert normalizePathEnd("lib", trailingSep = false) == "lib"
+        assert normalizePathEnd("/lib//", trailingSep = true) == "/lib/"
+    result = path
+    result.normalizePathEnd(trailingSep)
+
+else:
+  proc normalizePathEnd(path: var string, trailingSep = false) =
+    ## Ensures ``path`` has exactly 0 or 1 trailing `DirSep`, depending on
```
Comment (timotheecour) [inline lib/pure/os.nim:132]: i really don't like code duplication and we don't need runnableExamples + unittest for this un-exported version, all this is already in `normalizePathEnd*`

File: lib/pure/os.nim (hunk 1 above)
Comment (timotheecour) [reply lib/pure/os.nim:101]: so... how about using `since` ?

**Comment by timotheecour** [comment]
let's wait until https://github.com/nim-lang/Nim/pull/13123 is merged, then the simple:
```nim
when (NimMajor, NimMinor) >= (1, 1):
  export normalizePathEnd
```
would work, avoiding all the above issues

**Comment by timotheecour** [comment]
@disruptek https://github.com/nim-lang/Nim/pull/13123 got merged so you can now rebase against latest devel and use the simple
```
when (NimMajor, NimMinor) >= (1, 1):
  export normalizePathEnd
```

**Comment by Araq** [comment]
@timotheecour It still needs to be documented.

**Comment by timotheecour** [comment]
> @timotheecour It still needs to be documented.

this is what I mean: https://github.com/nim-lang/Nim/pull/13152 ; it shows documentation correctly and is simpler than this PR (and DRY)

**Comment by disruptek** [comment]
Nice.

**Pull request closed without merging**